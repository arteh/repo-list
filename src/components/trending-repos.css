.title {
  color: #FFFFFF;
}

.item {
/* background-color: rgb(75, 84, 59); */
  background-color: #F5F7DC;
  display: flex;
  margin: 10px 10px;
  padding: 10px 10px;
  border-radius: 5px;
  /* box-shadow: 0 0 10px rgb(169 169 169 / 70%); */

}


/* .item > span {
  flex: 0 0 25%;
  border-left: 2px solid #E65F5C;
} */

.favorite {
  flex: 0 0 10%;
}

.item-info {
  flex: 0 0 30%;
  border-right: 2px solid #E65F5C;
}

.item > span:first-child {
    border-left: none;
}

button {
  background: none;
  border: none;
  padding: 0;
  /* margin-right: auto; */
  cursor: pointer;
}

button > svg {
  width: 30px;
  height: 30px;
}
