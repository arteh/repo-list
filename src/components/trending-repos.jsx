import React, { useEffect, useState } from "react";
import { AiOutlineHeart, AiFillHeart, AiFillStar } from "react-icons/ai";

import "./list.css";

export const TrendingRepos = () => {
  const [repoList, setRepoList] = useState([]);
  const [faveList, setFaveList] = useState([]);

  // Get date 7 days before now
  const sevenDaysAgo = new Date(
    Date.now() - 7 * 24 * 60 * 60 * 1000
  ).toISOString();
  const formattedDate = sevenDaysAgo.substring(0, sevenDaysAgo.indexOf("T"));

  // Get data from github
  useEffect(() => {
    const url = `https://api.github.com/search/repositories?q=created:%3E${formattedDate}&sort=stars&order=desc`;

    const fetchData = async () => {
      try {
        const response = await fetch(url);
        const json = await response.json();
        console.log("json response", json);
        setRepoList(json.items);
      } catch (error) {
        console.log("error", error);
      }
    };
    fetchData();
  }, [formattedDate]);

  // Get favorite list from local storage
  useEffect(() => {
    let items = [],
      keys = Object.keys(localStorage).filter((key) =>
        key.includes("favorite")
      ),
      i = keys.length;

    while (i--) {
      items.push(JSON.parse(localStorage.getItem(keys[i])));
    }

    if (items) {
      setFaveList(items);
    }
  }, []);

  const handleFavorite = (item) => {
    localStorage.setItem(`favorite-${item.id}`, JSON.stringify(item));
  };

  return (
    <div>
      <h1>
        Most starred github repositories in the last 7 days
      </h1>

      {repoList ? (
        repoList?.map((item, index) => (
          <div className="item">
            <div className="name">{item?.name}</div>
            <div className="description">{item?.description}</div>
            <div className="stars">
              {item?.stargazers_count} <AiFillStar fill="#FFE046" />
            </div>
            <div className="favorite-btn">
              <button type="button" onClick={() => handleFavorite(item)}>
                {faveList.find((favorite) => favorite.id === item.id) ? (
                  <AiFillHeart fill="#EF8354" />
                ) : (
                  <AiOutlineHeart />
                )}
              </button>
            </div>
          </div>
        ))
      ) : (
        <p>Loading...</p>
      )}
    </div>
  );
};
