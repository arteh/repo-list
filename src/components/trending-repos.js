import React from "react";
import { AiOutlineHeart, AiFillHeart, AiOutlineStar } from "react-icons/ai";
import "./trending-repos.css";

export class TrendingRepos extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isDataReady: false,
      list: [],
      favoritesList: [],
    };
  }

  componentDidMount() {
      // move to helpers
    const sevenDaysAgo = new Date(
      Date.now() - 7 * 24 * 60 * 60 * 1000
    ).toISOString();
    const formattedDate = sevenDaysAgo.substring(0, sevenDaysAgo.indexOf("T"));

    fetch(
      `https://api.github.com/search/repositories?q=created:%3E${formattedDate}&sort=stars&order=desc`
    )
      .then((res) => res.json())
      .then((json) => {
        this.setState({
          list: json.items,
          isDataReady: true,
        });
      });
  }

  toggleFavorite = (item) => {
    console.log("item", item);

    const { favoritesList } = this.state;

    this.setState((prevState) => ({
      favoritesList: [...favoritesList, item],
    }));

    // TODO  we also need to do unfavourite
  };

  render() {
    const { isDataReady, list, favoritesList } = this.state;

    console.log("list in state", list);
    console.log("favoriteslist in state", favoritesList);

    return (
      <div>
        <h1 className="title">Most starred respoitories in the last 7 days</h1>
        {isDataReady ? (
          list?.map((item, index) => (
            <div className="item">
              <span className="item-info">{item?.name}</span>
              <span className="item-info">{item?.description}</span>
              <span className="item-info">{item?.stargazers_count} <AiOutlineStar /></span>
              <span className="favorite">
              <button
                type="button"
                value={item.id}
                onClick={() => this.toggleFavorite(item)}
              >
                {favoritesList.find((favorite) => favorite.id === item.id) ? (
                  <AiFillHeart fill="#E65F5C" />
                ) : (
                  <AiOutlineHeart />
                )}
              </button>
              </span>
     
            </div>
          ))
        ) : (
          <p>Loading...</p>
        )}
      </div>
    );
  }
}

// change to functional comp
// show the favourites in a different page
