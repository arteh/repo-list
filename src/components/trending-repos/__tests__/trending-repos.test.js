import React from "react";
import { screen, render, cleanup } from "@testing-library/react";
import { TrendingRepos } from "../trending-repos";

afterEach(cleanup);

describe("Trending Repos component", () => {
  it("should render the correct title", () => {
    render(<TrendingRepos />);
    expect(screen.getByText("Most starred github repositories in the last 7 days")).toBeInTheDocument();
  });

  it("should render the correct page elements", () => {
    const { asFragment } = render(<TrendingRepos />);
    expect(asFragment()).toMatchSnapshot();
  });
});
