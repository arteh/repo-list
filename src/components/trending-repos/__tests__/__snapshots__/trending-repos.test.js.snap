// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trending Repos component should render the correct page elements 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <h1>
      Most starred github repositories in the last 7 days
    </h1>
  </div>
</DocumentFragment>
`;
